spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/core/HealthCheckServiceInstanceListSupplier.java

```diff
@@ -56,50 +58,72 @@
 			LoadBalancerProperties.HealthCheck healthCheck, WebClient webClient) {
 		this.delegate = delegate;
 		this.healthCheck = healthCheck;
-		defaultHealthCheckPath = healthCheck.getPath().getOrDefault("default",
+		this.defaultHealthCheckPath = healthCheck.getPath().getOrDefault("default",
 				"/actuator/health");
 		this.webClient = webClient;
-		initInstances();
-
+		this.aliveInstancesReplay = Flux.defer(delegate)
+				.delaySubscription(Duration.ofMillis(this.healthCheck.getInitialDelay()))
+				.switchMap(serviceInstances -> healthCheckFlux(serviceInstances)
+						.map(alive -> Collections.unmodifiableList(new ArrayList<>(alive)))
+				)
+				.replay(1)
+				.refCount(1);
 	}
 
-	private void initInstances() {
-		delegate.get().subscribe(delegateInstances -> {
-			instances.clear();
-			instances.addAll(delegateInstances);
-		});
-
-		Flux<List<ServiceInstance>> healthCheckFlux = healthCheckFlux();
-
-		healthCheckFlux.subscribe(verifiedInstances -> {
-			healthyInstances.clear();
-			healthyInstances.addAll(verifiedInstances);
-		});
+	@Override
+	public void afterPropertiesSet() {
+		Disposable healthCheckDisposable = this.healthCheckDisposable;
+		if (healthCheckDisposable != null) {
+			healthCheckDisposable.dispose();
+		}
+		this.healthCheckDisposable = aliveInstancesReplay.subscribe();
 	}
 
-	protected Flux<List<ServiceInstance>> healthCheckFlux() {
-		return Flux.create(emitter -> Schedulers
-				.newSingle("Health Check Verifier: " + getServiceId(), true)
-				.schedulePeriodically(() -> {
-					List<ServiceInstance> verifiedInstances = new ArrayList<>();
-					Flux.fromIterable(instances).filterWhen(this::isAlive)
-							.subscribe(serviceInstance -> {
-								verifiedInstances.add(serviceInstance);
-								emitter.next(verifiedInstances);
-							});
-				}, healthCheck.getInitialDelay(), healthCheck.getInterval().toMillis(),
-						TimeUnit.MILLISECONDS),
-				FluxSink.OverflowStrategy.LATEST);
+	protected Flux<List<ServiceInstance>> healthCheckFlux(List<ServiceInstance> instances) {
+		return Flux.defer(() -> {
+			List<Mono<ServiceInstance>> checks = new ArrayList<>(instances.size());
+			for (ServiceInstance instance : instances) {
+				Mono<ServiceInstance> alive = isAlive(instance).onErrorResume(error -> {
+					if (LOG.isDebugEnabled()) {
+						LOG.debug(String.format(
+								"Exception occurred during health check of the instance for service %s: %s",
+								instance.getServiceId(), instance.getUri()), error);
+					}
+					return Mono.empty();
+				})
+				.timeout(this.healthCheck.getInterval(), Mono.defer(() -> {
+					if (LOG.isDebugEnabled()) {
+						LOG.debug(String.format(
+								"The instance for service %s: %s did not respond for %s during health check",
+								instance.getServiceId(), instance.getUri(),
+								this.healthCheck.getInterval()));
+					}
+					return Mono.empty();
+				}))
+				.handle((isHealthy, sink) -> {
```

Comment:
can be replaced with `.filter(isHealthy -> isHealthy)`

Reply:
@bsideup , not really.  
it can be replaced by  
```java
.filter(isHealthy -> isHealthy)
.map(it -> instance)
```
but is `filter` + `map` more efficient then `handle`? I thought `handle` is just for such cases